### ISSUE TITLE
Inference Time Variation

### ISSUE THREAD

**Issue by YASHSISODIA** [ultralytics/ultralytics#13550]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have converted the pre-trained yolov8n model(ie yolov8n.pt) specified on "[README.md](https://github.com/ultralytics/ultralytics/blob/main/README.md)" into openvino model using the below code

```
from ultralytics import YOLO
model = YOLO("yolov8n.pt")  
path = model.export(format="openvino", imgsz=[416,416])
```

Now i am using this openvino model for getting detections(bbox) on a video file.
Now what i have noticed is that
Case 1: When I run the model on every frame. The Inference time of the model is 20ms for 500 frames.
Case 2: When i use time.sleep(0.2) in between frames of the video. The Inference time of the model is 40ms for 500 frames.

Note: I am calculating the inference time like this in both the cases:

```
start_time = time.time()
res = exec_net.infer(inputs={input_blob: input_tensor})
end_time = time.time()
execution_time = end_time - start_time
total_time += execution_time
print(f"Execution time: {round(execution_time*1000)} seconds")
```

Also the detection result in both the cases are exactly same. Moreover i have tried the same on onnx model as well. The time without sleep and with sleep is approx 40ms and 80ms respectively.

I want to ask why their is so much time difference(approx 2 time) in the above case.
Please find below the python Script. You can reproduce the same by commenting and uncommenting the line time.sleep(0.2)

### Additional

```
import os
import cv2
import time
import numpy as np
from tqdm import tqdm
from openvino.inference_engine import IECore


def preprocess(image, input_shape, mean, std):
    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    resized = cv2.resize(image_rgb, (input_shape[1], input_shape[0]))

    # Scale input pixel value to 0 to 1
    input_image = resized / 255.0

    input_image = input_image - np.array(mean)
    input_image = input_image / np.array(std)

    # H*W*C --> C*H*W
    input_image = input_image.transpose(2, 0, 1)

    # B*C*H*W
    input_tensor = input_image[np.newaxis, :, :, :].astype(np.float32)

    return input_tensor


def get_detections(
    image, exec_net, input_blob, output_blob, input_shape, mean, std, conf_thresold
):
    image_height, image_width = image.shape[:2]
    input_height, input_width = input_shape[:2]

    input_tensor = preprocess(image, input_shape, mean, std)
    start_time = time.time()
    res = exec_net.infer(inputs={input_blob: input_tensor})
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Execution time: {round(execution_time*1000)} seconds")
    # Process the inference result
    outputs = res[output_blob]

    predictions = np.squeeze(outputs).T

    conf_thresold = conf_thresold
    # Filter out object confidence scores below threshold
    scores = np.max(predictions[:, 4:], axis=1)
    predictions = predictions[scores > conf_thresold, :]
    scores = scores[scores > conf_thresold]

    # Get the class with the highest confidence
    class_ids = np.argmax(predictions[:, 4:], axis=1)

    # Get bounding boxes for each object
    boxes = predictions[:, :4]

    # rescale box
    input_shape = np.array([input_width, input_height, input_width, input_height])
    boxes = np.divide(boxes, input_shape, dtype=np.float32)
    boxes *= np.array([image_width, image_height, image_width, image_height])
    boxes = boxes.astype(np.int32)

    return boxes, scores, class_ids


def xywh2xyxy(x):
    # Convert bounding box (x, y, w, h) to bounding box (x1, y1, x2, y2)
    y = np.copy(x)
    y[..., 0] = x[..., 0] - x[..., 2] / 2
    y[..., 1] = x[..., 1] - x[..., 3] / 2
    y[..., 2] = x[..., 0] + x[..., 2] / 2
    y[..., 3] = x[..., 1] + x[..., 3] / 2
    return y


def xyxy2xywh(x):
    # Convert bounding box (x1, y1, x2, y2) to bounding box (x, y, w, h)
    y = np.copy(x)
    y[..., 0] = x[..., 0] + (x[..., 2] - x[..., 0]) / 2
    y[..., 1] = x[..., 1] + (x[..., 3] - x[..., 1]) / 2
    y[..., 2] = x[..., 2] - x[..., 0]
    y[..., 3] = x[..., 3] - x[..., 1]
    return y


if __name__ == "__main__":
    save_annotations = True

    if save_annotations:
        save_predictions_score = True

    view_output_frames = False
    save_output_frames = True

    model_xml = "./best.xml"
    model_bin = "./best.bin"
    # input_dir = "/media/i2v-admin/2tb_drive/Raushan/wagon/yolo/images/val"
    output_dir = "./tmp2"

    conf_thresold = 0.4  # 0.25
    nms_threshold = 0.3  # 0.7

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    ie = IECore()

    net = ie.read_network(model=model_xml, weights=model_bin)

    input_blob = next(iter(net.input_info))
    output_blob = next(iter(net.outputs))

    exec_net = ie.load_network(network=net, device_name="CPU")

    input_shape = (416, 416)  # (H, W)
    mean = [0, 0, 0]
    std = [1, 1, 1]

    # classes = ["wagon_gap", "text"]
    classes = ["person", "bicycle", "car", "motorbike", "aeroplane", "bus", "train", "truck", "boat", 
		"traffic light", "fire hydrant", "stop sign", "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep", "cow", "elephant", "bear", "zebra",
		"giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee", "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove",
		"skateboard", "surfboard", "tennis racket", "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana", "apple", "sandwich", "orange", "broccoli",
		"carrot", "hot dog", "pizza", "donut", "cake", "chair", "sofa", "pottedplant", "bed", "diningtable", "toilet", "tvmonitor", "laptop", "mouse", "remote", "keyboard",
		"cell phone", "microwave", "oven", "toaster", "sink", "refrigerator", "book", "clock", "vase", "scissors", "teddy bear", "hair drier", "toothbrush"]

    # total_frames = sum([len(files) for root, dirs, files in os.walk(input_dir)])
    # pbar = tqdm(total=total_frames, unit="frame")

    frame_counter = -1

    video = cv2.VideoCapture("town.avi") 
    if (video.isOpened() == False):  
        print("Error reading video file") 
    
    videoHasData = True

    while(videoHasData):
    # for root, dirs, files in os.walk(input_dir):
        # for file in files:
        ret, frame = video.read() 
        if ret == True:  
            image = frame
            frame_counter += 1
            time.sleep(0.2)
            boxes, scores, class_ids = get_detections(
                image,
                exec_net,
                input_blob,
                output_blob,
                input_shape,
                mean,
                std,
                conf_thresold,
            )

            indices = cv2.dnn.NMSBoxes(
                boxes, scores, conf_thresold, nms_threshold
            )  # (bboxes, scores, score_threshold, nms_threshold)

            image_draw = image.copy()
            for bbox, score, label in zip(
                xywh2xyxy(boxes[indices]), scores[indices], class_ids[indices]
            ):
                bbox = bbox.round().astype(np.int32).tolist()
                cls_id = int(label)
                cls = classes[cls_id]
                color = (0, 255, 0)
                cv2.rectangle(image_draw, tuple(bbox[:2]), tuple(bbox[2:]), color, 2)
                cv2.putText(
                    image_draw,
                    f"{cls}:{int(score*100)}",
                    (bbox[0], bbox[1] + 30 - 2),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.60,
                    [0, 0, 255],
                    thickness=2,
                )

            if save_output_frames:
                out_dir = os.path.join(output_dir, "frames")
                if not os.path.exists(out_dir):
                    os.makedirs(out_dir)
                fileName = "f" + str(frame_counter) + ".jpg"
                cv2.imwrite(os.path.join(out_dir, fileName), image_draw)

            if view_output_frames:
                cv2.imshow("inference_result", image_draw)
                cv2.waitKey(1)

            # pbar.update(1)

        else:
            print("Video Ended")

    if view_output_frames:
        cv2.destroyAllWindows()
```

**Comment by glenn-jocher** [comment]
@YASHSISODIA hello,

Thank you for providing a detailed description and the reproducible code example. It's great to see that you've already explored the issue thoroughly. Let's address your query regarding the variation in inference time when using `time.sleep(0.2)` between frames.

### Explanation

The difference in inference time you're observing is likely due to how the CPU and other system resources are being managed. When you introduce a `time.sleep(0.2)` delay, the CPU may enter a low-power state or allocate resources differently, which can affect the performance of subsequent operations. This behavior is not specific to OpenVINO but is a general characteristic of how operating systems manage resources.

### Recommendations

1. **Consistent Benchmarking**: For consistent benchmarking, avoid introducing artificial delays like `time.sleep()`. Instead, measure the inference time in a continuous loop without interruptions.

2. **Warm-Up Iterations**: Include a few warm-up iterations before measuring inference time to ensure that the model and system are fully initialized and running at optimal performance.

3. **Resource Monitoring**: Monitor CPU and memory usage during inference to understand how system resources are being utilized. Tools like `htop` or `perf` can be helpful.

### Code Example without `time.sleep()`

Here's a modified version of your script that removes the `time.sleep(0.2)` and includes warm-up iterations:

```python
import os
import cv2
import time
import numpy as np
from openvino.inference_engine import IECore

def preprocess(image, input_shape, mean, std):
    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    resized = cv2.resize(image_rgb, (input_shape[1], input_shape[0]))
    input_image = resized / 255.0
    input_image = input_image - np.array(mean)
    input_image = input_image / np.array(std)
    input_image = input_image.transpose(2, 0, 1)
    input_tensor = input_image[np.newaxis, :, :, :].astype(np.float32)
    return input_tensor

def get_detections(image, exec_net, input_blob, output_blob, input_shape, mean, std, conf_thresold):
    input_tensor = preprocess(image, input_shape, mean, std)
    start_time = time.time()
    res = exec_net.infer(inputs={input_blob: input_tensor})
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Execution time: {round(execution_time * 1000)} ms")
    outputs = res[output_blob]
    predictions = np.squeeze(outputs).T
    scores = np.max(predictions[:, 4:], axis=1)
    predictions = predictions[scores > conf_thresold, :]
    scores = scores[scores > conf_thresold]
    class_ids = np.argmax(predictions[:, 4:], axis=1)
    boxes = predictions[:, :4]
    input_shape = np.array([input_shape[1], input_shape[0], input_shape[1], input_shape[0]])
    boxes = np.divide(boxes, input_shape, dtype=np.float32)
    boxes *= np.array([image.shape[1], image.shape[0], image.shape[1], image.shape[0]])
    boxes = boxes.astype(np.int32)
    return boxes, scores, class_ids

if __name__ == "__main__":
    model_xml = "./best.xml"
    model_bin = "./best.bin"
    output_dir = "./tmp2"
    conf_thresold = 0.4
    nms_threshold = 0.3

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    ie = IECore()
    net = ie.read_network(model=model_xml, weights=model_bin)
    input_blob = next(iter(net.input_info))
    output_blob = next(iter(net.outputs))
    exec_net = ie.load_network(network=net, device_name="CPU")
    input_shape = (416, 416)
    mean = [0, 0, 0]
    std = [1, 1, 1]

    video = cv2.VideoCapture("town.avi")
    if not video.isOpened():
        print("Error reading video file")

    # Warm-up iterations
    for _ in range(10):
        ret, frame = video.read()
        if not ret:
            break
        get_detections(frame, exec_net, input_blob, output_blob, input_shape, mean, std, conf_thresold)

    # Actual inference
    while video.isOpened():
        ret, frame = video.read()
        if not ret:
            break
        boxes, scores, class_ids = get_detections(frame, exec_net, input_blob, output_blob, input_shape, mean, std, conf_thresold)
        # Further processing...

    video.release()
    cv2.destroyAllWindows()
```

### Additional Resources

For more detailed information on optimizing and deploying YOLOv8 models with OpenVINO, please refer to the [OpenVINO Integration Guide](https://docs.ultralytics.com/integrations/openvino/).

I hope this helps! If you have any further questions or issues, feel free to ask.

**Comment by YASHSISODIA** [comment]
Thank you so much for your response. I genuinely appreciate the time and effort you took to reply. And I am grateful for your support and consideration. 

You were correct about the sleep method. Because when I replaced the sleep with an empty while loop running for 0.2 seconds, the inference time significantly decreased to 20ms.

So It seems that the increased inference time(40ms) is due to thread optimization. So I need to find a way to decrease the inference time while still using sleep or another method in my case. This is because I also want to reduce CPU usage when running more than one video stream. Optimizing this balance is crucial to ensure efficient performance while managing CPU load effectively. Any suggestions or insights on how to achieve this would be highly appreciated.

**Comment by glenn-jocher** [comment]
@YASHSISODIA hello,

Thank you for your kind words and for sharing your findings! It's great to hear that replacing the `sleep` method with an empty while loop helped reduce the inference time.

Given your requirement to balance inference time and CPU usage while managing multiple video streams, here are a few suggestions:

### 1. **Adjust Sleep Duration Dynamically**
Instead of a fixed `time.sleep(0.2)`, you could dynamically adjust the sleep duration based on the current CPU load. This way, you can optimize the balance between inference time and CPU usage.

### 2. **Use Asynchronous Inference**
OpenVINO supports asynchronous inference, which allows you to queue multiple inference requests and process them in parallel. This can help in managing multiple video streams more efficiently.

Here's an example of how to implement asynchronous inference:

```python
from openvino.inference_engine import IECore

# Initialize OpenVINO
ie = IECore()
net = ie.read_network(model=model_xml, weights=model_bin)
exec_net = ie.load_network(network=net, device_name="CPU")

# Asynchronous inference
request_id = 0
exec_net.start_async(request_id=request_id, inputs={input_blob: input_tensor})

# Wait for the result
status = exec_net.requests[request_id].wait()
if status == 0:
    res = exec_net.requests[request_id].outputs[output_blob]
```

### 3. **Thread Pooling**
Consider using a thread pool to manage multiple video streams. This can help in optimizing CPU usage by efficiently distributing the workload across multiple threads.

Here's a basic example using `concurrent.futures.ThreadPoolExecutor`:

```python
import concurrent.futures

def process_frame(frame):
    boxes, scores, class_ids = get_detections(
        frame, exec_net, input_blob, output_blob, input_shape, mean, std, conf_thresold
    )
    # Further processing...

with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
    while video.isOpened():
        ret, frame = video.read()
        if not ret:
            break
        executor.submit(process_frame, frame)
```

### 4. **CPU Affinity**
Set CPU affinity for your threads to ensure that they run on specific CPU cores. This can help in reducing context switching and improving performance.

```python
import os
import psutil

p = psutil.Process(os.getpid())
p.cpu_affinity([0, 1])  # Set affinity to CPU cores 0 and 1
```

### 5. **Optimize Model and Inference Engine**
Ensure that your OpenVINO model and inference engine are optimized for your specific hardware. You can use OpenVINO's Model Optimizer and Performance Hints to achieve this.

```python
import openvino.properties.hint as hints

config = {hints.performance_mode: hints.PerformanceMode.THROUGHPUT}
compiled_model = ie.compile_model(model, "CPU", config)
```

For more detailed information on optimizing OpenVINO inference, please refer to the [OpenVINO Optimization Guide](https://docs.ultralytics.com/guides/optimizing-openvino-latency-vs-throughput-modes/).

I hope these suggestions help you achieve the desired balance between inference time and CPU usage. If you have any further questions or need additional assistance, feel free to ask!
